test/e2e/pods.go

```diff
@@ -658,6 +658,364 @@ var _ = KubeDescribe("Pods", func() {
+		podClient := framework.Client.Pods(framework.Namespace.Name)
+
+		By("creating the pod")
+		name := "pod-init-" + string(util.NewUUID())
+		value := strconv.Itoa(time.Now().Nanosecond())
+		pod := &api.Pod{
+			ObjectMeta: api.ObjectMeta{
+				Name: name,
+				Labels: map[string]string{
+					"name": "foo",
+					"time": value,
+				},
+			},
+			Spec: api.PodSpec{
+				RestartPolicy: api.RestartPolicyNever,
+				InitContainers: []api.Container{
+					{
+						Name:    "init1",
+						Image:   "gcr.io/google_containers/busybox:1.24",
+						Command: []string{"/bin/true"},
+					},
+					{
+						Name:    "init2",
+						Image:   "gcr.io/google_containers/busybox:1.24",
+						Command: []string{"/bin/true"},
+					},
+				},
+				Containers: []api.Container{
+					{
+						Name:    "run1",
+						Image:   "gcr.io/google_containers/busybox:1.24",
+						Command: []string{"/bin/true"},
+					},
+				},
+			},
+		}
+		defer podClient.Delete(pod.Name, nil)
+		startedPod, err := podClient.Create(pod)
+		if err != nil {
+			Failf("Error creating a pod: %v", err)
+		}
+		w, err := podClient.Watch(api.SingleObject(startedPod.ObjectMeta))
+		if err != nil {
+			Failf("Error watching a pod: %v", err)
+		}
+		wr := watch.NewRecorder(w)
+		event, err := watch.Until(podStartTimeout, wr, client.PodCompleted)
+		expectNoError(err)
+		CheckInvariants(wr.Events(), ContainerInitInvariant)
+		endPod := event.Object.(*api.Pod)
+
+		Expect(endPod.Status.Phase).To(Equal(api.PodSucceeded))
+		init := api.GetPodCondition(endPod.Status, api.PodInitialized)
+		Expect(init).NotTo(BeNil())
+		Expect(init.Status).To(Equal(api.ConditionTrue))
+
+		Expect(len(endPod.Status.InitContainerStatuses)).To(Equal(2))
+		for _, status := range endPod.Status.InitContainerStatuses {
+			Expect(status.Ready).To(BeTrue())
+			Expect(status.State.Terminated).NotTo(BeNil())
+			Expect(status.State.Terminated.ExitCode).To(BeZero())
+		}
+	})
+
+	It("should invoke init containers on a RestartAlways pod [Conformance]", func() {
+		podClient := framework.Client.Pods(framework.Namespace.Name)
+
+		By("creating the pod")
+		name := "pod-init-" + string(util.NewUUID())
+		value := strconv.Itoa(time.Now().Nanosecond())
+		pod := &api.Pod{
+			ObjectMeta: api.ObjectMeta{
+				Name: name,
+				Labels: map[string]string{
+					"name": "foo",
+					"time": value,
+				},
+			},
+			Spec: api.PodSpec{
+				InitContainers: []api.Container{
+					{
+						Name:    "init1",
+						Image:   "gcr.io/google_containers/busybox:1.24",
+						Command: []string{"/bin/true"},
+					},
+					{
+						Name:    "init2",
+						Image:   "gcr.io/google_containers/busybox:1.24",
+						Command: []string{"/bin/true"},
+					},
+				},
+				Containers: []api.Container{
+					{
+						Name:  "nginx",
+						Image: "gcr.io/google_containers/pause:2.0",
+						Resources: api.ResourceRequirements{
+							Limits: api.ResourceList{
+								api.ResourceCPU:    *resource.NewMilliQuantity(100, resource.DecimalSI),
```

Comment:
Merge in the same test case (and loop?)  The behavior is different and the
test outcomes are slightly different, but probably.

On Fri, Apr 8, 2016 at 4:41 PM, Prashanth B <notifications@github.com>
wrote:

> In test/e2e/pods.go
> <https://github.com/kubernetes/kubernetes/pull/23567#discussion_r59086220>
> :
>
> > +						Image:   "gcr.io/google_containers/busybox:1.24",
> > +						Command: []string{"/bin/true"},
> > +					},
> > +					{
> > +						Name:    "init2",
> > +						Image:   "gcr.io/google_containers/busybox:1.24",
> > +						Command: []string{"/bin/true"},
> > +					},
> > +				},
> > +				Containers: []api.Container{
> > +					{
> > +						Name:  "nginx",
> > +						Image: "gcr.io/google_containers/pause:2.0",
> > +						Resources: api.ResourceRequirements{
> > +							Limits: api.ResourceList{
> > +								api.ResourceCPU:    *resource.NewMilliQuantity(100, resource.DecimalSI),
>
> Can you merge the RestartNever with RestartAlways tests?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23567/files/ad7a96fe3b9dda804fee77e41f7e14ad0b4513d7..98497e1265fe094c18542547514b0a96eee79c0a#r59086220>
>